Userland/Applications/PixelPaint/Image.cpp

```diff
@@ -379,38 +379,66 @@ ErrorOr<void> Image::merge_layers(LayerMergeMode layer_merge_mode)
-        return;
-    size_t layer_index = this->index_of(layer);
-    if ((layer_index + 1) == m_layers.size()) {
-        dbgln("Cannot merge layer up: layer is already at the top");
-        return; // FIXME: Notify user of error properly.
-    }
+    return merge_active_layer(layer, LayerMergeDirection::Up);
+}
 
-    auto& layer_above = m_layers.at(layer_index + 1);
-    GUI::Painter painter(layer_above.content_bitmap());
-    painter.draw_scaled_bitmap(rect(), layer.display_bitmap(), layer.rect(), (float)layer.opacity_percent() / 100.0f);
-    remove_layer(layer);
-    select_layer(&layer_above);
+ErrorOr<void> Image::merge_active_layer_down(Layer& layer)
+{
+    return merge_active_layer(layer, LayerMergeDirection::Down);
 }
 
-void Image::merge_active_layer_down(Layer& layer)
+ErrorOr<void> Image::merge_active_layer(Layer& layer, LayerMergeDirection layer_merge_direction)
 {
     if (m_layers.size() < 2)
-        return;
-    int layer_index = this->index_of(layer);
-    if (layer_index == 0) {
-        dbgln("Cannot merge layer down: layer is already at the bottom");
-        return; // FIXME: Notify user of error properly.
+        return {};
+
+    if (!layer.is_visible())
+        return Error::from_string_literal("Layer must be visible");
+
+    auto layer_index = index_of(layer);
+    auto direction = layer_merge_direction == LayerMergeDirection::Up ? 1 : -1;
+    ssize_t layer_to_merge_index = layer_index + direction;
+    ssize_t layer_count = m_layers.size();
+
+    if (layer_to_merge_index < 0)
+        return Error::from_string_literal("Layer is already at the bottom");
+    if (layer_to_merge_index >= layer_count)
+        return Error::from_string_literal("Layer is already at the top");
+
+    Optional<Layer&> maybe_adjacent_layer;
+    while (layer_to_merge_index >= 0 && layer_to_merge_index < layer_count) {
+        auto& layer = m_layers.at(layer_to_merge_index);
+        if (layer.is_visible()) {
+            maybe_adjacent_layer = layer;
+            break;
+        }
+        layer_to_merge_index += direction;
+    }
+
+    if (!maybe_adjacent_layer.has_value()) {
+        auto error_message = layer_merge_direction == LayerMergeDirection::Up ? "No visible layers above this layer"sv : "No visible layers below this layer"sv;
+        return Error::from_string_view(error_message);
     }
 
-    auto& layer_below = m_layers.at(layer_index - 1);
-    GUI::Painter painter(layer_below.content_bitmap());
-    painter.draw_scaled_bitmap(rect(), layer.display_bitmap(), layer.rect(), (float)layer.opacity_percent() / 100.0f);
-    remove_layer(layer);
-    select_layer(&layer_below);
+    auto& adjacent_layer = maybe_adjacent_layer.value();
+    auto merged_layer_bounding_rect = layer.relative_rect().united(adjacent_layer.relative_rect());
+    auto merged_layer = TRY(Layer::create_with_size(*this, merged_layer_bounding_rect.size(), adjacent_layer.name()));
+    merged_layer->set_location(merged_layer_bounding_rect.location());
```

Comment:
:thinking: If the upper layer is within the bounds of the lower layer, you could skip creating a new Layer object here, and just paint into the lower layer directly. That would make things more complicated though. No need to do it now, but maybe add a FIXME.

(The same is true of the merge-all-layers code in the previous commit.)

Reply:
`merge_layers`  and `merge_active_layer` now only create a new layer if the bottommost layer to be merged does not contain all other layers to be merged.